src/transform/src/join_implementation.rs

```diff
@@ -716,42 +707,64 @@ impl<'a> Orderer<'a> {
                     for expr in self.equivalences[*equivalence].iter() {
-                        if let ScalarExpr::Column(c) = expr {
-                            let rel = self.input_relation[*c];
-                            let col = *c - self.prior_arities[rel];
-
+                        // find the relations that columns in the expression belong to
+                        let rels = expr
+                            .support()
+                            .into_iter()
+                            .map(|c| self.input_relation[c])
+                            .collect::<Vec<_>>();
+                        // Skip the expression if
+                        // * the expression is a literal -> this would translate
+                        //   to `rels` being empty
+                        // * the expression has columns belonging to more than
+                        //   one relation -> TODO: see how we can plan better in
+                        //   this case. Arguably, if this happens, it would
+                        //   not be unreasonable to ask the user to write the
+                        //   query better.
+                        let rel = rels
+                            .get(0)
+                            .filter(|first_rel| rels[1..].iter().all(|rel| *first_rel == rel));
+                        if let Some(rel) = rel {
+                            let mut expr = expr.clone();
+                            expr.visit_mut(&mut |e| {
+                                if let ScalarExpr::Column(c) = e {
+                                    *c -= self.prior_arities[*rel];
+                                }
+                            });
                             // Update bound columns.
-                            self.bound[rel].push(col);
-                            self.bound[rel].sort();
+                            self.bound[*rel].push(expr);
+                            self.bound[*rel].sort();
+
                             // Reconsider all available arrangements.
-                            for (pos, keys) in self.arrangements[rel].iter().enumerate() {
-                                if !self.arrangement_active[rel].contains(&pos) {
+                            for (pos, keys) in self.arrangements[*rel].iter().enumerate() {
+                                if !self.arrangement_active[*rel].contains(&pos) {
                                     // Determine if the arrangement is viable, which happens when the
                                     // support of its keys are all bound.
-                                    if keys.iter().all(|k| {
-                                        k.support().iter().all(|c| self.bound[rel].contains(c))
-                                    }) {
-                                        self.arrangement_active[rel].push(pos);
+                                    if keys.iter().all(|k| self.bound[*rel].contains(k)) {
```

Comment:
This seems like a potential regression? For example, the previous test would allow an arrangement with key `a + b` as long as `a` and `b` are bound. This seems to require that `a + b` be bound itself.

I may have some logic backwards, but does this check out to you?

Reply:
Do you have an example of a query where you'd want the arrangement `a+b` to be allowed when only `a` and `b` are bound?

My thought was if you had
```
select *
from foo, bar
where foo.a = bar.a
and foo.b and bar.b
```
and an index
`foo(a + b)`

You wouldn't want the plan to become `Differential %1 %0(a+b)`

Reply:
Just to double check, if you mean
```
select *
from foo, bar
where foo.a = bar.a
and foo.b = bar.b
```
then an arrangement on `(a + b)` could in principle be useful, at least if we don't have one available on `(a, b)`. I'm not certain whether we should prefer building a new arrangement on `(a, b)` and then doing less post-filtering, or re-using an existing arrangement at the cost of CPU.

Maybe a better example is if `a: int4` is bound, should an arrangement on `to_int8(a)` become viable? It seems like it would make sense to do that, especially if we have that arrangement around and want to use it as part of a join with some `int8` thing.

This is all probably ahead of where we are now, but if we need to figure out which arrangements are available, I think this discovers fewer arrangements than before. That could be ok, but if it is in pursuit of allowing joins to use expression keys it seemed like it might interfere! At least, it's worth shaking out what we need here for expression-keyed arrangements.

Reply:
Currently on main, arrangements that have any expression columns are completely excluded from consideration (see line 138 in join_implementation.rs), and expression keys are not considered amongst the bounds. Thus, there would be no regression from the user perspective since currently no arrangements are being discovered when there is an expression key. 

Indeed the `and` in the example query was supposed to be an `=`. 

Right now if you give the render pipeline a plan that looks like
```
Join %0 %1 (= #0 #2) (=#1 #3)
| Implementation Differential %1 %0(#0 + #1)
```
The render pipeline will automatically create an index on `foo(a, b)` because we have programmed the render pipeline to look for an index that is an exact match of the join keys.

To make the arrangement on `(a+b)` useful, wouldn't we need to alter the plan to look something like this?
```
Join %0 %1 (= (#0 + #2) (#1 + #3))
| Implementation Differential %1 %0(a + b)
Filter (#0 = #2)
```
Which we could do some time in the future, but it seems complicated enough to be a completely new PR. 

Even in the `to_int8(a)` case, we would need to either
1) change the join key in the equivalence to `to_int8(a)` for the render pipeline to use the `to_int8(a)` index.
2) change dataflow rendering so it recognizes that `to_int8(a)` is substitutable for an index on `a`.

Reply:
Added a test that crashes in the dataflow rendering stage if `keys.iter().all(|k| self.bound[*rel].contains(k))` were to be changed to `keys.iter().all(|k| k.support().iter().all(|c| self.bound[rel].contains(c)))`. The crash occurs even though I added a filter that removes from consideration arrangements that contain any key that cannot also be found in in an equivalence.